### ISSUE TITLE
Anything imported hidden should be used

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/hlint#338]
If you `import Foo hiding (bar)` then `bar` should be used in the document. Check that GHC doesn't give an error on this first.

**Comment by ndmitchell** [comment]
Things not covered:

- `import Foo hiding (Bar)`
- `import Foo hiding (Bar(bar))`
- `import Foo hiding (Bar(Bar))`
- `import Foo hiding (pattern Bar)`
- `import Foo hiding (Bar(type Baz)`

**Comment by ndmitchell** [comment]
See #404 - it seems this hint doesn't really work, which is a shame, but hey ho.

**Comment by joneshf** [comment]
FWIW, I've found this hint to be incredibly useful. I wasn't bit by any of the cases that cause it to fail, but I'm curious if those cases could have been addressed with an `ANN` disabling the hint for the module.

If an `ANN` would work, would it be possible to have this hint again in a future version or is it too late?

**Comment by ndmitchell** [comment]
@joneshf I really found this incredibly useful too. I really like HLint to by default (in at least 99% of cases), so I'm happy to turn this hint back on (and would love to do so), but it has to have a lot less false positives - not just recommending people use `ANN`. Currently the best idea I have is that if you have import hiding (foo), and foo doesn't occur anywhere in the document, then the hiding is redundant. It doesn't catch the record wildcards+warning case, but perhaps that's crossing the 99% threshold? Alternative ideas welcome.

CC @mitchellwrosen

**Comment by mitchellwrosen** [comment]
The original code I wrote attempted to, given some manually hidden identifier, find all of the _uses_ of it. If there are none, complain that there's no reason to hide it.

However, as discovered in #404, there are more corner-cases than the original ones we thought of, which were basically just module re-exports:

- (1) RecordWildCards uses identifiers in `..`, which we can't know.
- (2) `main` is used in the `Main` module
- (3) Probably more :)

Apparently _shadowing_ an identifier was _not_ considered a "use". As in,

```haskell
import Foo hiding (blah)
blah = 5
```

The "unused warning" hint would fire here, complaining that there's no point in hiding `blah`, because it's not actually being used.

Perhaps that was wrong - if we _don't_ complain that `blah` is needlessly hidden in this example, that would cover both of the false positives that triggered `(1)` and `(2)` above.

**Comment by joneshf** [comment]
@ndmitchell That makes sense. Thanks for the insight.

> The "unused warning" hint would fire here, complaining that there's no point in hiding blah, because it's not actually being used.
> 
> Perhaps that was wrong - if we don't complain that blah is needlessly hidden in this example, that would cover both of the false positives that triggered (1) and (2) above.

That seems to make sense.

**Comment by joneshf** [comment]
I guess the question I have is whether it feels worthwhile at this point to try and re-support this hint. Looking through the other issues, it sounds like it hasn't been the easiest hint there is. 

Although I found it really helpful, it's not a deal breaker for me whether it exists or not. And I'd hate to put additional burden on you all to maintain it. It's totally reasonable to drop it if it seems like too much.

**Comment by ndmitchell** [comment]
I'm also concerned about the time it takes to check it, since scanning all variables in the file and building a Set of them is quite expensive. The maintenance is a lot less of a concern. I don't think I'm going to tackle this in the near future, but not against someone else sending a patch (probably best to look at the previous implementation to get inspiration/code).
